content/blog/sivayanama-testing-opsramp-apis-with-pytest-fixtures.md

```diff
@@ -0,0 +1,232 @@
+
+```python
+import http.client
+import json
+import traceback
+import urllib
+from asyncio.log import logger
+
+import pytest
+
+def get_auth_header(api_endpoint, client_id, client_secret, grant_type="client_credentials"):
+    """
+    Returns bearer token string in the blow format
+    bearer adfa-afdaf-1599-402c-a3ee-1ed24f597cc8
+    """
+
+    api_connection = http.client.HTTPSConnection(api_endpoint)
+    params = {"client_id": client_id, "client_secret": client_secret, "grant_type": "client_credentials"}
+
+    payload = urllib.parse.urlencode(params)
+
+    headers = {
+        'accept': "application/json",
+        'content-type': "application/x-www-form-urlencoded",
+    }
+
+```
+
+## Define Fixture to get Access Token
+
+You can transform the above function as PyTest Fixture by simply decorating the code as shown below 
+
+
+
+@pytest.fixture
+
+def get_auth_header(api_endpoint, client_id, client_secret, grant_type="client_credentials"):
+
+
+
+## Autouse fixtures
+
+Sometimes some of your fixtures are required for all other tests. The above Get access token is the perfect use case for this. Every API call is depending on this API. In this case you can make this fixture as “autouse fixture”  by passing in *autouse=True* to the fixture’s decorator. The autouse fixure capability will reduce lot of redundant requests. 
+
+```python
+@pytest.fixture(autouse=True)
+def get_auth_header(api_endpoint, client_id, client_secret, grant_type="client_credentials"):
+```
+
+
+
+## Provisioning of third-party plugin Fixtures 
+
+Not necessarily have to define all fixtures on your own code. Fixtures can be provided by third-party plugins as well and you are free to use them. You just need to install the required plugins. We will see an example with pytest-datadir plugin.
+
+
+
+Install this plugin as shown below 
+
+```shell
+pip install pytest-datadir
+```
+
+
+
+```python
+@pytest.fixture(autouse=True)
+
+def get_bearer_token(shared_datadir,request):
+
+json_path = (shared_datadir / "auth_header_input.json")
+```
+
+![](https://lh4.googleusercontent.com/dacTgDdw17BzeyCitShA73WSip9LVtenQoNN-uraaN5tKEU5cA_xP3cEmNPWmTzU3A1HegdoOVvwPbyYqQuoLeEk4W766nIvpBdoTzUdIiT2dXiOQG0_h7atQWS7-T9qvRrieuhlEV84VS15ir11Ocw)
+
+
+
+get\_bearer\_token fixture is using [shared_datadir](https://pypi.org/project/pytest-datadir/) fixture from third party plugin. Shared_datadir fixture provision data folder from the current folder as pathlib.Path object. 
+
+
+
+## Fixture finalization
+
+Usually, you do teardown or cleanup activities for each test for many good reasons such as the following: 
+
+- The executed tests will not impact other tests
+
+- Do not want to have the testing environment piled up with tons of test data
+
+- Every test should execute in a clean state
+
+
+
+Fixtures offer a very powerful teardown, cleanup capability known as finalization. 

```

Comment:
I'd recomment replacing the comma with a slash so it reads "... powerful teardown/cleanup capability..."